rpc/jsonrpc/server/ws_handler.go

```diff
@@ -284,14 +294,17 @@ func (wsc *wsConnection) Context() context.Context {
 
 // Read from the socket and subscribe to or unsubscribe from events
 func (wsc *wsConnection) readRoutine() {
+	// readRoutine will block until response is written
+	writeCtx := context.Background()
```

Comment:
So maybe we should use a timeout for it? Otherwise we can get goroutine leaks here.

Also is `readRoutine()` always called from one goroutine per connection? If not, we may want to do the panic write in a separate goroutine, to avoid blocking.

Reply:
> So maybe we should use a timeout for it? Otherwise we can get goroutine leaks here.

updated comment. WriteRPCResponse exits if WS connection is closed (I don't expect any leaks here).

> Also is readRoutine() always called from one goroutine per connection?

yes, there can only be 1 reader and 1 writer per WS connection (as stated in gorilla/websocket docs)